src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs

```diff
@@ -122,17 +123,34 @@ internal class OnAutoInsertHandler : IRequestHandler<LSP.DocumentOnAutoInsertPar
             }
 
             var (service, context) = serviceAndContext.Value;
-            var postReturnEdit = await service.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: false, cancellationToken).ConfigureAwait(false);
+            var postReturnEdit = await service.GetTextChangeAfterReturnAsync(context, cancellationToken).ConfigureAwait(false);
             if (postReturnEdit == null)
             {
                 return null;
             }
 
-            var textChange = await GetCollapsedChangeAsync(postReturnEdit.Value, document, cancellationToken).ConfigureAwait(false);
-            Debug.Assert(postReturnEdit.Value.CaretLocation >= textChange.Span.Start);
-            var offsetInTextChange = postReturnEdit.Value.CaretLocation - textChange.Span.Start;
+            var textChanges = postReturnEdit.Value.TextChanges;
+            var desiredCaretLinePosition = postReturnEdit.Value.CaretLocation;
+            var newSourceText = sourceText.WithChanges(textChanges);
+
+            var caretLine = newSourceText.Lines[desiredCaretLinePosition.Line];
+            if (caretLine.IsEmptyOrWhitespace() && desiredCaretLinePosition.Character > caretLine.Span.Length)
```

Comment:
i'm not opposed to this.  but part of me feels things would just be simpler if we only handled the 'empty' case.  handling the whitespace case makes it feel like you're likely writing code that we shouldn't be able to hit.

Reply:
other things to be resilient for (or assert can't happen): what if it's a non-empty-non-whitespace line and the Character is still past the end?

i think it would make sense to potentially clamp the Character to the line, unless the line is purely empty.  might be simpler and easier to undestand/maintain.